source/common/http/conn_manager_impl.cc

```diff
@@ -260,6 +260,15 @@ void ConnectionManagerImpl::doDeferredStreamDestroy(ActiveStream& stream) {
 
   read_callbacks_->connection().dispatcher().deferredDelete(stream.removeFromList(streams_));
 
+  if (stream.response_encoder_) {
+    stream.response_encoder_->getStream().removeCallbacks(stream);
```

Comment:
Where was this done previously or was it not done at all? Can you add more comments on why we are doing this here?

Reply:
It wasn't being done at all e.g. it was a dangling listener to stream callbacks. So if any event fired after this is deferred destroyed, this could cause a crash (as the added integration test shows via a reset occuring after the Envoy level stream is destroyed). This is the originating crash call stack you're seeking in https://github.com/envoyproxy/envoy/pull/18895/files/0b98a9c3e2b14f85cc50f696ea84779829cc1caf#r751419027

Now by unsubscribing,  the Envoy level stream from the subject codec level stream, unveiled another problem where in the H1 local reply case, the codec level stream can be destroyed before the Envoy level stream.

In the H1 local reply case, `stream.response_encoder` would be a dangling pointer in here, but `onEnvoyStreamComplete()` helps us by defer the deletion of the codec until we're done with it (unsubscribing Active stream).

Reply:
Sorry for being dense, but can you provide an actual backtrace of the crash without the h1 fix? I'm just trying to understand what the crash looks like exactly. It's a bit hard for me to understand that from the above diagram.

Reply:
> as the added integration test shows via a reset occuring after the Envoy level stream is destroyed

Separate question: As dicey as all of this code is, I'm skeptical that we would not have already heard reports of crashes if a reset came in due to held buffer in the codec. (AIUI, the codec would not raise reset if both local/remote are ended). Is there something about the overload manager that is causing this issue to crop up?

Reply:
I'm actually a fan of removing the callbacks, but if we're going to do that why do we need a separate onEnvoyStreamComplete?  I'd assume the lack of callbacks would be suffiicent to deduce that?

Reply:
Here the stack trace of the original crash from the integration test: 

```
[ RUN      ] IpVersions/Http2OverloadManagerIntegrationTest.CanResetStreamIfEnvoyLevelStreamEnded/IPv4_Http2Downstream_Http2UpstreamBareHttp2_with_per_stream_buffer_accounting_BareHttp2
[2021-11-17 18:01:44.075][1998][critical][assert] [./source/common/stream_info/stream_info_impl.h:124] assert failure: !final_time_.
[2021-11-17 18:01:44.075][1998][critical][backtrace] [./source/server/backtrace.h:104] Caught Aborted, suspect faulting address 0x9456a0000000c
[2021-11-17 18:01:44.075][1998][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2021-11-17 18:01:44.075][1998][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 0/1.21.0-dev/test/DEBUG/BoringSSL
[2021-11-17 18:01:44.143][1998][critical][backtrace] [./source/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x4d1fa99]
[2021-11-17 18:01:44.144][1998][critical][backtrace] [./source/server/backtrace.h:96] #1: __restore_rt [0x7fa21a2328e0]
[2021-11-17 18:01:44.209][1998][critical][backtrace] [./source/server/backtrace.h:96] #2: Envoy::Http::ConnectionManagerImpl::ActiveStream::completeRequest() [0x39cce42]
[2021-11-17 18:01:44.274][1998][critical][backtrace] [./source/server/backtrace.h:96] #3: Envoy::Http::ConnectionManagerImpl::doDeferredStreamDestroy() [0x39ccc7d]
[2021-11-17 18:01:44.338][1998][critical][backtrace] [./source/server/backtrace.h:96] #4: Envoy::Http::ConnectionManagerImpl::ActiveStream::onResetStream() [0x39d2d61]
[2021-11-17 18:01:44.403][1998][critical][backtrace] [./source/server/backtrace.h:96] #5: Envoy::Http::StreamCallbackHelper::runResetCallbacks() [0x40d77de]
[2021-11-17 18:01:44.468][1998][critical][backtrace] [./source/server/backtrace.h:96] #6: Envoy::Http::Http2::ConnectionImpl::StreamImpl::resetStream() [0x40f2feb]
[2021-11-17 18:01:44.533][1998][critical][backtrace] [./source/server/backtrace.h:96] #7: Envoy::Http::Http2::ConnectionImpl::ServerStreamImpl::resetStream() [0x40f2f90]
[2021-11-17 18:01:44.598][1998][critical][backtrace] [./source/server/backtrace.h:96] #8: Envoy::Buffer::BufferMemoryAccountImpl::resetDownstream() [0x4ad7cf3]
[2021-11-17 18:01:44.663][1998][critical][backtrace] [./source/server/backtrace.h:96] #9: Envoy::Buffer::WatermarkBufferFactory::resetAccountsGivenPressure() [0x4ad4d5f]
[2021-11-17 18:01:44.727][1998][critical][backtrace] [./source/server/backtrace.h:96] #10: Envoy::Server::WorkerImpl::resetStreamsUsingExcessiveMemory() [0x3020b9f]
[2021-11-17 18:01:44.792][1998][critical][backtrace] [./source/server/backtrace.h:96] #11: Envoy::Server::WorkerImpl::WorkerImpl()::$_2::operator()() [0x3021708]
[2021-11-17 18:01:44.856][1998][critical][backtrace] [./source/server/backtrace.h:96] #12: std::__invoke_impl<>() [0x30216c9]
[2021-11-17 18:01:44.920][1998][critical][backtrace] [./source/server/backtrace.h:96] #13: std::__invoke_r<>() [0x3021652]
[2021-11-17 18:01:44.984][1998][critical][backtrace] [./source/server/backtrace.h:96] #14: std::_Function_handler<>::_M_invoke() [0x3021542]
[2021-11-17 18:01:45.051][1998][critical][backtrace] [./source/server/backtrace.h:96] #15: std::function<>::operator()() [0x369456c]
[2021-11-17 18:01:45.118][1998][critical][backtrace] [./source/server/backtrace.h:96] #16: Envoy::Server::OverloadManagerImpl::flushResourceUpdates()::$_5::operator()() [0x366cd5d]
[2021-11-17 18:01:45.184][1998][critical][backtrace] [./source/server/backtrace.h:96] #17: std::__invoke_impl<>() [0x366cd0d]
[2021-11-17 18:01:45.249][1998][critical][backtrace] [./source/server/backtrace.h:96] #18: std::__invoke_r<>() [0x366ccad]
[2021-11-17 18:01:45.314][1998][critical][backtrace] [./source/server/backtrace.h:96] #19: std::_Function_handler<>::_M_invoke() [0x366cbad]
[2021-11-17 18:01:45.379][1998][critical][backtrace] [./source/server/backtrace.h:96] #20: std::function<>::operator()() [0x289b8ee]
[2021-11-17 18:01:45.444][1998][critical][backtrace] [./source/server/backtrace.h:96] #21: Envoy::Event::DispatcherImpl::runPostCallbacks() [0x4a85a07]
[2021-11-17 18:01:45.509][1998][critical][backtrace] [./source/server/backtrace.h:96] #22: Envoy::Event::DispatcherImpl::DispatcherImpl()::$_3::operator()() [0x4a89d98]
[2021-11-17 18:01:45.574][1998][critical][backtrace] [./source/server/backtrace.h:96] #23: std::__invoke_impl<>() [0x4a89d5d]
[2021-11-17 18:01:45.639][1998][critical][backtrace] [./source/server/backtrace.h:96] #24: std::__invoke_r<>() [0x4a89cfd]
[2021-11-17 18:01:45.705][1998][critical][backtrace] [./source/server/backtrace.h:96] #25: std::_Function_handler<>::_M_invoke() [0x4a89bfd]
[2021-11-17 18:01:45.705][1998][critical][backtrace] [./source/server/backtrace.h:96] #26: std::function<>::operator()() [0x289b8ee]
[2021-11-17 18:01:45.770][1998][critical][backtrace] [./source/server/backtrace.h:96] #27: Envoy::Event::SchedulableCallbackImpl::SchedulableCallbackImpl()::$_0::operator()() [0x4cd1ba1]
[2021-11-17 18:01:45.835][1998][critical][backtrace] [./source/server/backtrace.h:96] #28: Envoy::Event::SchedulableCallbackImpl::SchedulableCallbackImpl()::$_0::__invoke() [0x4cd1b66]
[2021-11-17 18:01:45.900][1998][critical][backtrace] [./source/server/backtrace.h:96] #29: event_process_active_single_queue [0x50a2066]
[2021-11-17 18:01:45.966][1998][critical][backtrace] [./source/server/backtrace.h:96] #30: event_process_active [0x509c75a]
[2021-11-17 18:01:46.030][1998][critical][backtrace] [./source/server/backtrace.h:96] #31: event_base_loop [0x509b60c]
[2021-11-17 18:01:46.101][1998][critical][backtrace] [./source/server/backtrace.h:96] #32: Envoy::Event::LibeventScheduler::run() [0x4ccf184]
[2021-11-17 18:01:46.168][1998][critical][backtrace] [./source/server/backtrace.h:96] #33: Envoy::Event::DispatcherImpl::run() [0x4a8550a]
[2021-11-17 18:01:46.233][1998][critical][backtrace] [./source/server/backtrace.h:96] #34: Envoy::Server::WorkerImpl::threadRoutine() [0x3020658]
[2021-11-17 18:01:46.299][1998][critical][backtrace] [./source/server/backtrace.h:96] #35: Envoy::Server::WorkerImpl::start()::$_6::operator()() [0x3022aa5]
[2021-11-17 18:01:46.364][1998][critical][backtrace] [./source/server/backtrace.h:96] #36: std::__invoke_impl<>() [0x3022a5d]
[2021-11-17 18:01:46.429][1998][critical][backtrace] [./source/server/backtrace.h:96] #37: std::__invoke_r<>() [0x3022a0d]
[2021-11-17 18:01:46.494][1998][critical][backtrace] [./source/server/backtrace.h:96] #38: std::_Function_handler<>::_M_invoke() [0x302285d]
[2021-11-17 18:01:46.494][1998][critical][backtrace] [./source/server/backtrace.h:96] #39: std::function<>::operator()() [0x289b8ee]
[2021-11-17 18:01:46.560][1998][critical][backtrace] [./source/server/backtrace.h:96] #40: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::operator()() [0x5080a02]
[2021-11-17 18:01:46.625][1998][critical][backtrace] [./source/server/backtrace.h:96] #41: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::__invoke() [0x50809d5]
[2021-11-17 18:01:46.626][1998][critical][backtrace] [./source/server/backtrace.h:96] #42: start_thread [0x7fa21a227eae]
```

I think the issue the OM is pointing out, is we still have a registered dangling observer. The only other reset I believe that can occur after the Envoy level stream is destroyed is the pending flush timer (with only uses a low level reset mechanism via this timer https://github.com/envoyproxy/envoy/blob/41438c95a92be7ef24b6c389fc3a93f6fbedd7fb/source/common/http/codec_helper.h#L101). The OM would ideally would use the highest level reset available (e.g. b/c better for stats / observability). If we just used the resetStreamWorker all the time, we'd avoid the dangling listener b/c we wouldn't fire a callback, but the great observability we try to provide wouldn't be there.

Reply:
> I'm actually a fan of removing the callbacks, but if we're going to do that why do we need a separate onEnvoyStreamComplete? I'd assume the lack of callbacks would be suffiicent to deduce that?

It's for the issue with the codec stream not living longer than the envoy stream as discussed above, (relevant parts brought below): 
Now by unsubscribing, the Envoy level stream from the subject codec level stream, unveiled another problem where in the H1 local reply case, the codec level stream can be destroyed before the Envoy level stream.

In the H1 local reply case, stream.response_encoder would be a dangling pointer in here, but onEnvoyStreamComplete() helps us by defer the deletion of the codec until we're done with it (unsubscribing Active stream).

Reply:
OK I understand the integration test crash and why we haven't seen it before, it's because it's a local reset vs. a remote reset.

For the H1 crash that happens without the additional fix, what is the call stack of that crash in the local reply case?

Reply:
So just patching in the:
```
 if (stream.response_encoder_) {
    stream.response_encoder_->getStream().removeCallbacks(stream);
    stream.response_encoder_ = nullptr;
}
```
here.
We get H1 crashes such as the following from protocol_integration_test:

```
[ RUN      ] Protocols/DownstreamProtocolIntegrationTest.MissingHeadersLocalReply/IPv6_HttpDownstream_HttpUpstreamBareHttp2
[2021-11-18 01:05:34.145][83][critical][backtrace] [./source/server/backtrace.h:104] Caught Segmentation fault, suspect faulting address 0x5c228180d6a8
[2021-11-18 01:05:34.145][83][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2021-11-18 01:05:34.145][83][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 0/1.21.0-dev/test/DEBUG/BoringSSL
[2021-11-18 01:05:34.216][83][critical][backtrace] [./source/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x4dffb19]
[2021-11-18 01:05:34.216][83][critical][backtrace] [./source/server/backtrace.h:96] #1: __restore_rt [0x7f80bb45f8e0]
[2021-11-18 01:05:34.286][83][critical][backtrace] [./source/server/backtrace.h:96] #2: Envoy::Http::ConnectionManagerImpl::doEndStream() [0x3e60f0d]
[2021-11-18 01:05:34.356][83][critical][backtrace] [./source/server/backtrace.h:96] #3: Envoy::Http::ConnectionManagerImpl::ActiveStream::endStream() [0x3e81dda]
[2021-11-18 01:05:34.425][83][critical][backtrace] [./source/server/backtrace.h:96] #4: Envoy::Http::FilterManager::maybeEndEncode() [0x3eb5e1a]
[2021-11-18 01:05:34.495][83][critical][backtrace] [./source/server/backtrace.h:96] #5: Envoy::Http::FilterManager::encodeHeaders() [0x3eae5ae]
[2021-11-18 01:05:34.565][83][critical][backtrace] [./source/server/backtrace.h:96] #6: Envoy::Http::FilterManager::sendLocalReplyViaFilterChain()::$_9::operator()() [0x3ebd499]
[2021-11-18 01:05:34.635][83][critical][backtrace] [./source/server/backtrace.h:96] #7: std::__invoke_impl<>() [0x3ebd3ec]
[2021-11-18 01:05:34.706][83][critical][backtrace] [./source/server/backtrace.h:96] #8: std::__invoke_r<>() [0x3ebd377]
[2021-11-18 01:05:34.776][83][critical][backtrace] [./source/server/backtrace.h:96] #9: std::_Function_handler<>::_M_invoke() [0x3ebd1b7]
[2021-11-18 01:05:34.846][83][critical][backtrace] [./source/server/backtrace.h:96] #10: std::function<>::operator()() [0x3099fc2]
[2021-11-18 01:05:34.915][83][critical][backtrace] [./source/server/backtrace.h:96] #11: Envoy::Http::Utility::sendLocalReply() [0x4d6daa5]
[2021-11-18 01:05:34.985][83][critical][backtrace] [./source/server/backtrace.h:96] #12: Envoy::Http::FilterManager::sendLocalReplyViaFilterChain() [0x3eb4c5d]
[2021-11-18 01:05:35.055][83][critical][backtrace] [./source/server/backtrace.h:96] #13: Envoy::Http::FilterManager::sendLocalReply() [0x3eaa0d7]
[2021-11-18 01:05:35.126][83][critical][backtrace] [./source/server/backtrace.h:96] #14: Envoy::Http::ActiveStreamDecoderFilter::sendLocalReply() [0x3ea941e]
[2021-11-18 01:05:35.196][83][critical][backtrace] [./source/server/backtrace.h:96] #15: Envoy::InvalidHeaderFilter::decodeHeaders() [0x4ab168e]
[2021-11-18 01:05:35.266][83][critical][backtrace] [./source/server/backtrace.h:96] #16: Envoy::Http::ActiveStreamDecoderFilter::decodeHeaders() [0x3ec2545]
[2021-11-18 01:05:35.336][83][critical][backtrace] [./source/server/backtrace.h:96] #17: Envoy::Http::FilterManager::decodeHeaders() [0x3ea2563]
[2021-11-18 01:05:35.406][83][critical][backtrace] [./source/server/backtrace.h:96] #18: Envoy::Http::FilterManager::decodeHeaders() [0x3e7f966]
[2021-11-18 01:05:35.475][83][critical][backtrace] [./source/server/backtrace.h:96] #19: Envoy::Http::ConnectionManagerImpl::ActiveStream::decodeHeaders() [0x3e6fc5e]
[2021-11-18 01:05:35.544][83][critical][backtrace] [./source/server/backtrace.h:96] #20: Envoy::Http::Http1::ServerConnectionImpl::onMessageCompleteBase() [0x41d293c]
[2021-11-18 01:05:35.613][83][critical][backtrace] [./source/server/backtrace.h:96] #21: Envoy::Http::Http1::ConnectionImpl::onMessageComplete() [0x41cd362]
[2021-11-18 01:05:35.683][83][critical][backtrace] [./source/server/backtrace.h:96] #22: Envoy::Http::Http1::LegacyHttpParserImpl::Impl::Impl()::{lambda()#3}::operator()() [0x41e9d87]
[2021-11-18 01:05:35.752][83][critical][backtrace] [./source/server/backtrace.h:96] #23: Envoy::Http::Http1::LegacyHttpParserImpl::Impl::Impl()::{lambda()#3}::__invoke() [0x41e9d45]
[2021-11-18 01:05:35.822][83][critical][backtrace] [./source/server/backtrace.h:96] #24: http_parser_execute [0x4dbd280]
[2021-11-18 01:05:35.892][83][critical][backtrace] [./source/server/backtrace.h:96] #25: Envoy::Http::Http1::LegacyHttpParserImpl::Impl::execute() [0x41e8fad]
[2021-11-18 01:05:35.961][83][critical][backtrace] [./source/server/backtrace.h:96] #26: Envoy::Http::Http1::LegacyHttpParserImpl::execute() [0x41e8612]
[2021-11-18 01:05:36.031][83][critical][backtrace] [./source/server/backtrace.h:96] #27: Envoy::Http::Http1::ConnectionImpl::dispatchSlice() [0x41c7258]
[2021-11-18 01:05:36.101][83][critical][backtrace] [./source/server/backtrace.h:96] #28: Envoy::Http::Http1::ConnectionImpl::dispatch() [0x41c55f9]
[2021-11-18 01:05:36.170][83][critical][backtrace] [./source/server/backtrace.h:96] #29: Envoy::Http::Http1::ConnectionImpl::dispatch() [0x41c843b]
[2021-11-18 01:05:36.239][83][critical][backtrace] [./source/server/backtrace.h:96] #30: Envoy::Http::ConnectionManagerImpl::onData() [0x3e65b41]
[2021-11-18 01:05:36.309][83][critical][backtrace] [./source/server/backtrace.h:96] #31: Envoy::Network::FilterManagerImpl::onContinueReading() [0x4bac5f1]
[2021-11-18 01:05:36.378][83][critical][backtrace] [./source/server/backtrace.h:96] #32: Envoy::Network::FilterManagerImpl::onRead() [0x4baca9a]
[2021-11-18 01:05:36.448][83][critical][backtrace] [./source/server/backtrace.h:96] #33: Envoy::Network::ConnectionImpl::onRead() [0x4b94a4f]
[2021-11-18 01:05:36.517][83][critical][backtrace] [./source/server/backtrace.h:96] #34: Envoy::Network::ConnectionImpl::onReadReady() [0x4b9fed9]
[2021-11-18 01:05:36.586][83][critical][backtrace] [./source/server/backtrace.h:96] #35: Envoy::Network::ConnectionImpl::onFileEvent() [0x4b9bcd9]
[2021-11-18 01:05:36.655][83][critical][backtrace] [./source/server/backtrace.h:96] #36: Envoy::Network::ConnectionImpl::ConnectionImpl()::$_6::operator()() [0x4ba664e]
[2021-11-18 01:05:36.725][83][critical][backtrace] [./source/server/backtrace.h:96] #37: std::__invoke_impl<>() [0x4ba6611]
[2021-11-18 01:05:36.795][83][critical][backtrace] [./source/server/backtrace.h:96] #38: std::__invoke_r<>() [0x4ba65a2]
[2021-11-18 01:05:36.864][83][critical][backtrace] [./source/server/backtrace.h:96] #39: std::_Function_handler<>::_M_invoke() [0x4ba6492]
[2021-11-18 01:05:36.934][83][critical][backtrace] [./source/server/backtrace.h:96] #40: std::function<>::operator()() [0x2e8a38a]
[2021-11-18 01:05:37.004][83][critical][backtrace] [./source/server/backtrace.h:96] #41: Envoy::Event::DispatcherImpl::createFileEvent()::$_5::operator()() [0x4b6a374]
[2021-11-18 01:05:37.075][83][critical][backtrace] [./source/server/backtrace.h:96] #42: std::__invoke_impl<>() [0x4b6a321]
[2021-11-18 01:05:37.145][83][critical][backtrace] [./source/server/backtrace.h:96] #43: std::__invoke_r<>() [0x4b6a2c2]
[2021-11-18 01:05:37.214][83][critical][backtrace] [./source/server/backtrace.h:96] #44: std::_Function_handler<>::_M_invoke() [0x4b6a142]
[2021-11-18 01:05:37.214][83][critical][backtrace] [./source/server/backtrace.h:96] #45: std::function<>::operator()() [0x2e8a38a]
[2021-11-18 01:05:37.284][83][critical][backtrace] [./source/server/backtrace.h:96] #46: Envoy::Event::FileEventImpl::mergeInjectedEventsAndRunCb() [0x4b78bf0]
[2021-11-18 01:05:37.354][83][critical][backtrace] [./source/server/backtrace.h:96] #47: Envoy::Event::FileEventImpl::assignEvents()::$_1::operator()() [0x4b79119]
[2021-11-18 01:05:37.422][83][critical][backtrace] [./source/server/backtrace.h:96] #48: Envoy::Event::FileEventImpl::assignEvents()::$_1::__invoke() [0x4b78c66]
[2021-11-18 01:05:37.492][83][critical][backtrace] [./source/server/backtrace.h:96] #49: event_persist_closure [0x5182b0e]
[2021-11-18 01:05:37.561][83][critical][backtrace] [./source/server/backtrace.h:96] #50: event_process_active_single_queue [0x5182138]
[2021-11-18 01:05:37.631][83][critical][backtrace] [./source/server/backtrace.h:96] #51: event_process_active [0x517c8fa]
[2021-11-18 01:05:37.700][83][critical][backtrace] [./source/server/backtrace.h:96] #52: event_base_loop [0x517b7ac]
[2021-11-18 01:05:37.770][83][critical][backtrace] [./source/server/backtrace.h:96] #53: Envoy::Event::LibeventScheduler::run() [0x4daf204]
[2021-11-18 01:05:37.840][83][critical][backtrace] [./source/server/backtrace.h:96] #54: Envoy::Event::DispatcherImpl::run() [0x4b648da]
[2021-11-18 01:05:37.910][83][critical][backtrace] [./source/server/backtrace.h:96] #55: Envoy::Server::WorkerImpl::threadRoutine() [0x34a2108]
[2021-11-18 01:05:37.981][83][critical][backtrace] [./source/server/backtrace.h:96] #56: Envoy::Server::WorkerImpl::start()::$_6::operator()() [0x34a4555]
[2021-11-18 01:05:38.052][83][critical][backtrace] [./source/server/backtrace.h:96] #57: std::__invoke_impl<>() [0x34a450d]
[2021-11-18 01:05:38.121][83][critical][backtrace] [./source/server/backtrace.h:96] #58: std::__invoke_r<>() [0x34a44bd]
[2021-11-18 01:05:38.187][83][critical][backtrace] [./source/server/backtrace.h:96] #59: std::_Function_handler<>::_M_invoke() [0x34a430d]
[2021-11-18 01:05:38.252][83][critical][backtrace] [./source/server/backtrace.h:96] #60: std::function<>::operator()() [0x29b74ce]
[2021-11-18 01:05:38.318][83][critical][backtrace] [./source/server/backtrace.h:96] #61: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::operator()() [0x5160b62]
[2021-11-18 01:05:38.383][83][critical][backtrace] [./source/server/backtrace.h:96] #62: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::__invoke() [0x5160b35]
[2021-11-18 01:05:38.383][83][critical][backtrace] [./source/server/backtrace.h:96] #63: start_thread [0x7f80bb454eae]
ActiveStream 0x5c227d62f500, stream_id_: 1346260036310749978&filter_manager_:
  FilterManager 0x5c227d62f578, state_.has_1xx_headers_: 0
  filter_manager_callbacks_.requestHeaders():
    ':authority', 'host'
    ':path', '/test/long/url'
    ':scheme', 'http'
    'remove-method', 'yes'
    'send-reply', 'yes'
    'x-forwarded-proto', 'http'
    'x-request-id', '759eb464-1cfb-4769-b9fa-e5153cdff0e4'
  filter_manager_callbacks_.requestTrailers():   null
  filter_manager_callbacks_.responseHeaders():
    ':status', '200'
    'date', 'Thu, 18 Nov 2021 01:05:34 GMT'
    'server', 'envoy'
  filter_manager_callbacks_.responseTrailers():   null
  &stream_info_:
    StreamInfoImpl 0x5c227d62f678, upstream_connection_id_: null, protocol_: 1, response_code_: 200, response_code_details_: InvalidHeaderFilter_ready, attempt_count_: null, health_check_request_: 0, route_name_:
    OverridableRemoteConnectionInfoSetterStreamInfo 0x5c227d62f678, remoteAddress(): [::1]:50840, directRemoteAddress(): [::1]:50840, localAddress(): [::1]:38829
Http1::ConnectionImpl 0x5c227d62fc08, dispatching_: 1, dispatching_slice_already_drained_: 0, reset_stream_called_: 0, handling_upgrade_: 0, deferred_end_stream_headers_: 1, require_strict_1xx_and_204_headers_: 1, send_strict_1xx_and_204_headers_: 1, processing_trailers_: 0, no_chunked_encoding_header_for_304_: 1, buffered_body_.length(): 0, header_parsing_state_: Done, current_header_field_: , current_header_value_:
active_request_: null

absl::get<RequestHeaderMapPtr>(headers_or_trailers_): null
current_dispatching_buffer_ front_slice length: 80 contents: "GET /test/long/url HTTP/1.1\r\nhost: host\r\nremove-method: yes\r\nsend-reply: yes\r\n\r\n"
ConnectionImpl 0x5c227d60a5c0, connecting_: 0, bind_error_: 0, state(): Open, read_buffer_limit_: 1048576
socket_:
  ListenSocketImpl 0x5c227d8c4200, transport_protocol_: raw_buffer
  connection_info_provider_:
    ConnectionInfoSetterImpl 0x5c227d8c2250, remote_address_: [::1]:50840, direct_remote_address_: [::1]:50840, local_address_: [::1]:38829, server_name_:

```

I think we end up crashing when we try to deference the `response_encoder_` because it's a dangling pointer at this point. Looking at the tracked object stack printout we see that `active_request_: null` which holds the object we pointed to using `response_encoder_`. I think that gets reset due to 

https://github.com/envoyproxy/envoy/blob/aac4631c26745de1a39556abc68bd1d3ed95aa2f/source/common/http/http1/codec_impl.cc#L1005-L1012

Reply:
Can you run it under the debugger and see exactly what line is crashing? I'm kind of confused now.

The if statement in the function above should prevent active_request_ from getting deleted so it can wait around for the reset. It seems like something else must be broken? I'm hoping we can do a smaller fix?

Reply:
Specifically I'm wondering if we are missing a `state_.codec_saw_local_complete_ = true;` somewhere in the local reply path. cc @alyssawilk @snowp

Reply:
Sorry, maybe I misunderstood the ask, the patch I talked about in https://github.com/envoyproxy/envoy/pull/18895#discussion_r751809850  is **excluding** the `stream.response_encoder_->getStream().onEnvoyStreamComplete();` (and related logic), in order to demonstrate the reason for having the `onEnvoyStreamComplete()` as it helps defer deletion of the H1 stream in a local reply scenario until after we've unsubscribed from listening.

Running it **without** the onEnvoyStramComplete logic in gdb, here the output:

```
[ RUN      ] Protocols/DownstreamProtocolIntegrationTest.MissingHeadersLocalReply/IPv4_HttpDownstream_HttpUpstreamBareHttp2
[New Thread 0x7ffff7c30640 (LWP 2099330)]
[New Thread 0x7ffff742f640 (LWP 2099331)]
[New Thread 0x7ffff6c1e640 (LWP 2099332)]
[New Thread 0x7ffff641d640 (LWP 2099333)]
[New Thread 0x7ffff5c1c640 (LWP 2099334)]
[New Thread 0x7ffff541b640 (LWP 2099335)]
[New Thread 0x7ffff4c0a640 (LWP 2099336)]
[New Thread 0x7ffff4409640 (LWP 2099337)]
[New Thread 0x7ffff3c08640 (LWP 2099338)]
[New Thread 0x7ffff3407640 (LWP 2099339)]
[New Thread 0x7ffff2c06640 (LWP 2099340)]

Thread 10 "wrk:worker_0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff3c08640 (LWP 2099338)]
0x0000000003e62281 in Envoy::Http::ConnectionManagerImpl::doDeferredStreamDestroy (this=0x4c6d3dd84560, stream=...)
    at source/common/http/conn_manager_impl.cc:264
264         stream.response_encoder_->getStream().removeCallbacks(stream);
(gdb) bt
#0  0x0000000003e62281 in Envoy::Http::ConnectionManagerImpl::doDeferredStreamDestroy (this=0x4c6d3dd84560, stream=...)
    at source/common/http/conn_manager_impl.cc:264
#1  0x0000000003e61f8d in Envoy::Http::ConnectionManagerImpl::doEndStream (this=0x4c6d3dd84560, stream=...)
    at source/common/http/conn_manager_impl.cc:230
#2  0x0000000003e82e5a in Envoy::Http::ConnectionManagerImpl::ActiveStream::endStream (this=0x4c6d3ddcb500)
    at ./source/common/http/conn_manager_impl.h:261

```

Reply:
Ah OK. It's crashing on the new call. OK I understand. I'm thinking and will report back.

Reply:
OK thanks for the back and forth here. I really appreciate you taking the time to explain this issue fully and get me up to speed on this.

Given what I know now, I'm still generally not in favor of this fix mainly because it makes things even harder to understand (at least from my perspective). A few thoughts for possible other solutions. Let me know what you think.
1) Can you defer delete the H1 active request? I'm pretty sure this will fix the issue and would also parallel what the H2 codec does (which is probably why that codec doesn't crash in the same flow)
2) If (1) doesn't work, another path forward would be to slightly refactor the HCM code so it's more clear when the HCM expects the stream to be gone, and remove callbacks *before* that point. So for example, in the remote complete / local complete case, before encoding to the codec the final frame that will cause the stream to go away, remove callbacks. In the reset flows we would keep callbacks to allow the local reset callbacks if any to flow.

Reply:
Thanks for taking the time to get up to speed on this. 

I think that the defer deletion of the H1 active request will fix the problem as you suggest in (1).  We can have it follow how h2 streams are deferred deleted. So the active request should still be alive (although added to deferred deletion queue), when we go to remove the Envoy level stream as an observer of the encoder it holds. Since these (local reply and destroying the Envoy level stream) occur as part of the same callback, the deferred deletion queue shouldn’t be cleared in between.

If that sounds good to you, I’ll go ahead and implement that.

Reply:
Yup SGTM, thank you.

Reply:
Since I conflated a few fixes in the PR (making it potentially hard to understand), I've broken out the refactoring HTTP1 Active Request to be deferred deletable. The code looks much cleaner: https://github.com/envoyproxy/envoy/pull/19062